### ISSUE TITLE
ITML accuracy worse than guessing

### ISSUE THREAD

**Issue by drjosephliu** [scikit-learn-contrib/metric-learn#286]
I'm using the [Omniglot](https://github.com/brendenlake/omniglot) dataset, which has 20 samples of 964 classes of images, each which is 1 x 105 x 105.

I'm embedding these samples down to 512 dimensions. So the final dataset has shape `(964*20, 512)`.

To implement a 5-way one-shot task, for each one of the 964 classes, I create a support set of five images; the first one is another sample of the same class and the other four are from different classes. The tuples therefore consist of the query image with one of the support set images. 

This is what it looks like below (I'm using preprocessor with indices):
```
# For each index, make a 5-way 1-shot task with four other randomly selected classes
indices = []
n_classes, n_examples, dim = train_embeddings.reshape(-1, 20, 512).shape
for i in range(n_classes):
    
    ex1 = rng.randint(n_examples)
    ex2 = rng.randint(n_examples)
    indices.append([i* 20 + ex1, i * 20 + ex2]) # First pair is from the same class
    
    # Remaining four pairs are from different classes
    for j in range(4):
        random_class = rng.randint(n_classes)
        while random_class == i:
            random_class = rng.randint(n_classes)
        ex3 = rng.randint(n_examples)
        indices.append([i * 20 + ex1, random_class* 20 + ex3])

labels = [1,-1,-1,-1,-1] * n_classes
indices, labels = shuffle(indices, labels)

itml = ITML(preprocessor=train_embeddings)
itml.fit(indices, labels)
```
When it comes to testing time, instead of using `predict()`, I use instead the `score_pairs()` function and return the index with the highest score. If it corresponds to the same index in the gold labels, then I return that as a correctly classified task:

```
pair_scores = model.score_pairs(encoded_pairs)
if np.argmax(pair_scores) == np.argmax(targets):
     return 1
```

With all this in mind, I'm getting a 5-way accuracy of 0%. 3-way comes out to 12% though so it's not completely aberrant, but still way worse than random guessing. Am I using this algorithm correctly?

**Comment by wdevazelhes** [comment]
Dear @drjosephliu , sorry for the late answer, and thank you for reporting this interesting example

I am not very familiar with one-shot learning, so I'm not sure to understand the testing in your setting: you have a reference sample and 5 (or 3) other samples (one of which is from the same class as the reference class), and the goal is to select the one of these samples that is the same from the reference is that right ? 

If so, did you check the generalization of the model using cross-validation on pairs for instance, for selecting the hyperparameters ? Maybe you can also add more samples by redoing your trainset preparation several times for the same class ? Also maybe you can try to balance the dataset of pairs by having as much positive as negative pairs, although I'm not sure what effect if would have…

Otherwise, it is also possible that even so, ITML performs poorly, in which case I don't really know… Maybe it is not adapted to the task, or maybe our implementations have some problems…

I hope this helps

**Comment by wdevazelhes** [comment]
@drjosephliu EDIT:  in the end, I realized the `score_pairs` function returns the distance between the two samples of the pairs, so I guess you should rather check the minimum, not the maximum score
http://contrib.scikit-learn.org/metric-learn/generated/metric_learn.ITML.html?highlight=itml#metric_learn.ITML.score_pairs
I hope this helps

**Comment by bellet** [comment]
It is true that the term `score` is a bit misleading. But you should really use `decision_function`, which follows the convention of sklearn (the larger, the higher the probability of being positive)
